### PR TITLE
Dev attribution info

### PR DESCRIPTION
**Pull request by EnBakir**
A component that shows information about the attribution of the basemap(s).

### CODE REVIEW

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		const translate = (key) => this._translationService.translate(key);
+		const { active } = this._state;
+
+		if (typeof active !== undefined && active.length > 0) {
+			const geoResource = this._georesourceService.byId(active[0].id);
+			if (!geoResource) {
+				return nothing;
+			} 
+			// Not yet implemented
+			geoResource.attribution ? this._content = geoResource.attribution : this._content = 'No data available';
```
Comment (thiloSchlemmer) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:34]: 'No data available' is good candidate for the TranslationService, i guess.

File: test/modules/footer/components/attributionInfo/AttributionInfo.test.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+			const wmts = new WMTSGeoResource('someId', 'LDBV42', 'https://some{1-2}/layer/{z}/{x}/{y}');
+			const geoServiceMock = spyOn(geoResourceServiceMock, 'byId').withArgs(layer.id).and.returnValue(wmts);
+
+			const element = await setup(state);
+
+			// Not yet implemented so the fallback content is shown
+			expect(element.shadowRoot.querySelector('p')).toBeTruthy();
+			expect(element.shadowRoot.querySelector('p').innerHTML).toContain('map_attributionInfo_label');
+			expect(element.shadowRoot.querySelector('p').innerHTML).toContain('No data available');
```
Comment (thiloSchlemmer) [inline test/modules/footer/components/attributionInfo/AttributionInfo.test.js:42]: Albeit the attribution-property is not already implemented and a valid content structure could change...i would prefer, to have a test for a simple but valid attribution-content (layer.attribution !== null -> layer.attribution='FooBarBaz' should be OK for now).

**Review by thiloSchlemmer — CHANGES_REQUESTED** [review]
See inline comments

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * @author bakir_en 
+ */
+export class AttributionInfo extends BaElement {
+
+	constructor() {
+		super();
+		const { TranslationService, GeoResourceService } = $injector.inject('TranslationService', 'GeoResourceService');
+		this._translationService = TranslationService;
+		this._georesourceService = GeoResourceService;
+		this._content = null;
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:17]: Avoid unnecessary private fields, which makes your code more resistant to unwanted side-effects

**Review by taulinger — COMMENTED** [review]
Well on track!
You could also already provide the zoom-level information from the store, we will need it for zoom-level dependent attributions.

**Review by taulinger — COMMENTED** [review]
Nice, I will extend the georesource domain classes now.
In the meantime you could provide the zoom-level information from the store, we will need it for zoom-level dependent attributions.

**Review by thiloSchlemmer — APPROVED** [review]

**Comment by taulinger** [comment]
@EnBakir Added the attribution definition and its access in `GeoResources`. Please have look at the attribution's `@typedef` doc to get an idea about guaranteed and optionally values

**Comment by EnBakir** [comment]
The `AttributionInfo.js` is now using the `georesource.getAttribution(zoom)` Method. Next steps are: 

1. Refactoring: Move `attributionInfo` to `map/components` folder
2. Extend `BaseLayerInfo.js`

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 4]
```diff
@@ -22,15 +22,29 @@ export class AttributionInfo extends BaElement {
...
 				return nothing;
 			} 
-			// Not yet implemented
-			geoResource.attribution ? this._content = geoResource.attribution : this._content = translate('map_attributionInfo_fallback');
+
+			const attributions = geoResource.getAttribution(zoom);
+
+			if (!attributions) {
+				return nothing;
+			}
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:41]: Have a look @ attribution typedef in `geoResources.js`.
An attribution object MAY also contain an `url` property. In this case, we need an anchor here.

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 5]
```diff
@@ -38,24 +39,43 @@ export class AttributionInfo extends BaElement {
...
+					} 
+				}
+				else if (attribution.copyright.label != null) {
+					if (index < attributionsLength - 1) {
+						attributionCopyright.push(html`<p class='attribution-label'> ${attribution.copyright.label}, </p>`);
+					}
+					else {
+						attributionCopyright.push(html`<p class='attribution-label'> ${attribution.copyright.label} </p>`);
+					} 
+				}
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:63]: The only difference in each case is the appended semicolon. Can we reduce the redundant code here (-> DRY)?

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+
+	/**
+     * @override 
+     */
+	createView() {
+		const translate = (key) => this._translationService.translate(key);
+		const { active, zoom } = this._state;
+
+		if (typeof active !== undefined && active.length > 0) {
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:28]: Unneeded check: `active` is guaranteed not to be null or undefined by the layers reducer

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+						attributionCopyright.push(html`<a class='attribution-link' target='new' href=${attribution.copyright.url} > ${attribution.copyright.label} </a>`);
+					} 
+				}
+				else if (attribution.copyright.label != null) {
+					if (index < attributionsLength - 1) {
+						attributionCopyright.push(html`<p class='attribution-label'> ${attribution.copyright.label}, </p>`);
+					}
+					else {
+						attributionCopyright.push(html`<p class='attribution-label'> ${attribution.copyright.label} </p>`);
+					}
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:62]: Do we really need `<p>` elements in this case?

File: src/modules/footer/components/attributionInfo/AttributionInfo.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+					} 
+				} 
+			});
+			
+			// At least the first element should not be null
+			if (attributionCopyright[0] === null || attributionCopyright[0] === undefined) {
+				return html`
+            		<div><p>${translate('map_attributionInfo_fallback')}</p></div>
+				`;
+			}
```
Comment (taulinger) [inline src/modules/footer/components/attributionInfo/AttributionInfo.js:71]: No need to check for an empty array. `geoResource.getAttribution()` makes sure you will either get `null` or an array with at least one element

File: test/modules/footer/components/attributionInfo/AttributionInfo.test.js [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+			const element = await setup(state);			
+
+			expect(element.shadowRoot.querySelector('p').innerHTML).toContain('map_attributionInfo_fallback');
+
+			expect(geoServiceMock).toHaveBeenCalledOnceWith(layer.id);
+			expect(getAttrMock).toHaveBeenCalledOnceWith(12);
+		});
+
+		it('updates BaseLayerInfo component', async ()  => {
```
Comment (taulinger) [inline test/modules/footer/components/attributionInfo/AttributionInfo.test.js:133]: wrong label (BaseLayerInfo) also in line 183

File: src/modules/map/components/attributionInfo/AttributionInfo.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+
+	/**
+     * @override 
+     */
+	createView() {
+		const translate = (key) => this._translationService.translate(key);
+		const { active, zoom } = this._state;
+
+		const geoResource = active[0] ?  this._georesourceService.byId(active[0].id) : null;
```
Comment (taulinger) [inline src/modules/map/components/attributionInfo/AttributionInfo.js:28]: Sorry, did not see this the last time:
In contrast to the base layer info (where we take the current base layer only), here we have to display the attribution for **all active AND visible layers**

**Review by alsturm — APPROVED** [review]
well done! added style, moved it to footer.

**Comment by taulinger** [comment]
@alsturm Can we improve the layout for "long" attributions? 
For instance:
localhost:8080?l=atkis,9d0e3859-be17-4a40-b439-1ba19b45fbb8,5d4af972-fa72-48e0-a8c1-55d0782e540a,d8f6e9a6-d5ef-4234-a77c-e6253f279e57,26d2b2b8-3944-4a49-aec2-59f827d9aa9e,a818e80e-78b7-4b21-8c15-d5f8abe6c420

File: src/modules/map/components/attributionInfo/AttributionInfo.js [hunk 11]
```diff
@@ -25,32 +25,47 @@ export class AttributionInfo extends BaElement {
...
+		}
 
-		if (!attributions) {
+
+		// render fallback if no layers are set or visible
+		if (attributionsRaw[0] === null || attributionsRaw[0] === undefined) {
 			return html`
             		<div>${translate('map_attributionInfo_fallback')}</div>
 				`;
 		}
```
Comment (taulinger) [inline src/modules/map/components/attributionInfo/AttributionInfo.js]: I think we can remove the complete fallback case now, it's not needed anymore...

**Review by taulinger — COMMENTED** [review]
Let's remove the fallback (see inline comment)

**Review by taulinger — COMMENTED** [review]
Thx, let's wait for @alsturm  UI optimizations...
Maybe we have to add additional functionality like shortening the attribution string and
a toggle function. We will have a look at this next week....

**Review by taulinger — COMMENTED** [review]
@alsturm Nice proceedings
- Can we improve the formatting of line break (see screenshot)
For instance:
localhost:8080?l=atkis,9d0e3859-be17-4a40-b439-1ba19b45fbb8,5d4af972-fa72-48e0-a8c1-55d0782e540a,d8f6e9a6-d5ef-4234-a77c-e6253f279e57,26d2b2b8-3944-4a49-aec2-59f827d9aa9e,a818e80e-78b7-4b21-8c15-d5f8abe6c420

- if we show a title for opening the full content, we also need one for closing.... maybe we just name it "show all" and "close"

**Review by taulinger — APPROVED** [review]
Looks good now 👍

**Comment by taulinger** [comment]
Let's merge it ...

**Pull request merged**